worker-taskunstowing/src/main/java/com/github/jobservice/workers/taskunstowing/TaskUnstowingWorker.java

```diff
@@ -0,0 +1,195 @@
+            return;
+        }
+
+        final String jobId = document.getCustomData("jobId");
+        if (Strings.isNullOrEmpty(jobId)) {
+            final TaskUnstowingWorkerFailure failure = new TaskUnstowingWorkerFailure(
+                TaskUnstowingWorkerFailure.JOB_ID_MISSING_FROM_CUSTOM_DATA_ID,
+                "Custom data should contain a non-empty 'jobId' property");
+            processFailure(failure, document);
+            return;
+        }
+
+        final WorkerTaskData workerTaskData = document.getTask().getService(WorkerTaskData.class);
+        if (workerTaskData == null) {
+            final TaskUnstowingWorkerFailure failure = new TaskUnstowingWorkerFailure(
+                TaskUnstowingWorkerFailure.FAILED_TO_GET_WORKER_TASK_DATA_ID,
+                "Failed to get worker task data");
+            processFailure(failure, document);
+            return;
+        }
+
+        LOGGER.info("Querying for stowed tasks for partition id {} and job id {}", partitionId, jobId);
+        final List<StowedTaskRow> stowedTaskRows;
+        try {
+            stowedTaskRows = databaseClient.getStowedTasks(partitionId, jobId);
+            LOGGER.info("Found {} stowed task(s) for partition id {} and job id {}", stowedTaskRows.size(), partitionId, jobId);
+        } catch (final Exception exception) {
+            final TaskUnstowingWorkerFailure failure = new TaskUnstowingWorkerFailure(
+                TaskUnstowingWorkerFailure.FAILED_TO_READ_FROM_DATABASE_ID,
+                String.format("Failed to read stowed task(s) for partition id %s and job id %s from database", partitionId, jobId));
+            processFailure(failure, exception, document);
+            return;
+        }
+
+        for (final StowedTaskRow stowedTaskRow : stowedTaskRows) {
+            final TaskMessage taskMessage;
+            try {
+                taskMessage = convertStowedTaskRowToTaskMessage(stowedTaskRow);
+            } catch (final IOException exception) {
+                final TaskUnstowingWorkerFailure failure = new TaskUnstowingWorkerFailure(
+                    TaskUnstowingWorkerFailure.FAILED_TO_CONVERT_DATABASE_ROW_TO_TASK_MESSAGE_ID,
+                    String.format("Failed to convert database row with partition id {}, job id {} and job task id {} to task message",
+                                  stowedTaskRow.getPartitionId(), stowedTaskRow.getJobId(), stowedTaskRow.getTrackingInfoJobTaskId()));
+                processFailure(failure, exception, document);
+                continue;
+            }
+
+            try {
+                workerTaskData.sendMessage(taskMessage);
+                LOGGER.info("Sent unstowed task message with partition id {}, job id {} and job task id {} to queue {}",
+                            partitionId, jobId, taskMessage.getTracking().getJobTaskId(), taskMessage.getTo());
+                try {
+                    databaseClient.deleteStowedTask(partitionId, jobId);
```

Comment:
But this is going to delete all the tasks associated with the job, not just the one that has just been dispatched.  This is a serious issue.  If the worker goes down immediately after this then it will lead to data loss.  Even if it was done after the loop it could still be a problem if it deleted extra messages that have been stowed since the rows were retrieved in line 102.

Reply:
I've updated the delete query to take a jobTaskId as well now to only delete that particular row.

Reply:
Ok but that will mean that we need a database change as well, so that rows can be looked up by (job id, task id).

Reply:
Ah yes ok, I have already merged the table into job service:

https://github.com/JobService/job-service/pull/241

Can I revert the commit or will that break things?

Reply:
No need to revert it; just put it a further update.

Reply:
I'm looking up by(partition id, job id, job task id) in the delete query, so was going to create an index on those 3. You mention (job id, job task id) above though?

Reply:
The 3-part index is fine - covers both scenarios in one index.